src/assert.js

```diff
@@ -479,13 +370,98 @@ class Assert {
 					expected,
 					message
 				} );
-
 				done();
 			}
 		);
 	}
 }
 
+function validateExpectedExceptionArgs( expected, message, assertionMethod ) {
+	const expectedType = objectType( expected );
+
+	// 'expected' is optional unless doing string comparison
+	if ( expectedType === "string" ) {
+		if ( message === undefined ) {
+			message = expected;
+			expected = undefined;
+			return [ expected, message ];
+		} else {
+			throw new Error(
+				"assert." + assertionMethod +
+				" does not accept a string value for the expected argument.\n" +
+				"Use a non-string object value (e.g. RegExp or validator function) " +
+				"instead if necessary."
+			);
+		}
+	}
+
+	const valid =
+		expected === undefined ||
+		expected === null ||
+		expectedType === "regexp" ||
+		expectedType === "function" ||
+		expectedType === "object";
+
+	if ( !valid ) {
+		const message =
+			"Invalid expected value type (" + expectedType + ") " +
+			"provided to assert." + assertionMethod + ".";
+		throw new Error( message );
+	}
+
+	return [ expected, message ];
+}
+
+function validateException( actual, expected, message ) {
+	let result = false;
+	const expectedType = objectType( expected );
+
+	// These branches should be exhaustive, based on validation done in validateExpectedException
+
+	// We don't want to validate
+	if ( expected === undefined || expected === null ) {
```

Comment:
I haven't confirmed it, but it seems from reading the removed code, that previously a `null` expected value would reach the `else` branch that reports "invalid expected value provided". If so, did you mean to change that?

I didn't actually realize that null (non-object and unboxable primitive) was valid in a throw statement, but it looks like it actually is possible to throw `null` (the same way strings and numbers can be thrown).

I'm neutral on whether we need to support that, but I think the current state might be a bit strange, to have `null` behave the same as no argument, e.g. when would one use `assert.throws(fn, null)` instead of `assert.throws(fn)`?

Reply:
The old `throws` logic actually used a `!expected` check
([assert.js#L325](https://github.com/qunitjs/qunit/blob/main/src/assert.js#L325)), so undefined, null, false, 0, etc could all make it through here to be "valid" with no matching at all, just that it threw.

The old `rejects` logic used a more explicit `expected === undefined` check ([assert.js#L435](https://github.com/qunitjs/qunit/blob/main/src/assert.js#L435)), so the null/false/0 sort of values would be considered invalid and produce failures. That made for some inconsistencies between the two.

To target the "nullish" values, I went with `expected === undefined || expected === null`. So I've restricted the null/false/0 values for both (which now throw, not fail), but now a `null` expected value for `rejects` has changed from an invalid type to a valid/ignored one. I could go one further and restrict it to _only_ `undefined`, so that a `null` value would error as well for both.

What do you think?

Reply:
I think any value other than `undefined` should either be considered invalid, or be used as comparison target in some way. Using `null` in this way would be unexpected, in my opinion. I did not realize that `assert.throws()` was implicitly allowing null and other falsely values as if they are undefined. We probably shouldn't change that in a minor release.

Basically, `undefined` is here to represent the case of no `expected` parameter being passed and asserting only that something was thrown. A stricter version of this code would have used `arguments.length` to detect that, instead of comparing to undefined.

Two ideas:
1. In this commit, change both to use `!expected` (assuming empty string remains invalid and handled earlier). That should not cause any previously passing code to start failing, and will temporarily lax `rejects()` to start tolerating any other garbage passed to it. Assumign that people's existing tests are passing, this should not change any behaviour, but will add a temporary blind splot. Then in 3.0 we could change it to  `=== undefined` (or `arguments.length`) and consider unhandled types as invalid.
2. Alternatively, perhaps vary this section of the code by `method` such that the behaviour remains the same for both. The throws branch accepts falsey as alias for undefined, the rejects branch considers non-undefined falsey values as invalid. The throws branch can then be deprecated in a follow-up commit.